libpromises/evalfunction.c

```diff
@@ -736,16 +736,35 @@ static FnCallResult FnCallHostsSeen(ARG_UNUSED EvalContext *ctx, ARG_UNUSED cons
 }
 
 /*********************************************************************/
+enum {
+    HOSTSWITHCLASS,
+    HOSTSWITHGROUP
+};
 
-static FnCallResult FnCallHostsWithClass(EvalContext *ctx, ARG_UNUSED const Policy *policy, ARG_UNUSED const FnCall *fp, const Rlist *finalargs)
+static FnCallResult FnCallHostsWithField(EvalContext *ctx, ARG_UNUSED const Policy *policy, ARG_UNUSED const FnCall *fp, const Rlist *finalargs)
 {
+    assert(fp != NULL);
     Rlist *returnlist = NULL;
 
     char *class_name = RlistScalarValue(finalargs);
     char *return_format = RlistScalarValue(finalargs->next);
-
-    if (!ListHostsWithClass(ctx, &returnlist, class_name, return_format))
+    if (StringEqual(fp->name, "hostswithclass"))
+    {
+        if (!ListHostsWithField(ctx, &returnlist, class_name, return_format, HOSTSWITHCLASS))
+        {
+            return FnFailure();
+        }
+    }
+    else if (StringEqual(fp->name, "hostswithgroup"))
+    {
+        if (!ListHostsWithField(ctx, &returnlist, class_name, return_format, HOSTSWITHGROUP))
+        {
+            return FnFailure();
+        }
+    }
+    else
     {
+        Log(LOG_LEVEL_ERR, "HostswithField with unknown call function %s, aborting", fp->name);
```

Comment:
I would use the `ProgrammingError()` function here. Because this is not a user error and means that there is something seriously wrong with the code.
```suggestion
        Log(LOG_LEVEL_ERR, "HostsWithField with unknown call function %s, aborting", fp->name);
```